packages/node-core/src/indexer/poi/poi.service.ts

```diff
@@ -1,55 +1,261 @@
+          await this.syncPoiJob(poiBlocks, logging);
+        }
+        if (poiBlocks.length < DEFAULT_FETCH_RANGE) {
+          await delay(POI_AWAIT_TIME);
+        }
       }
     }
-    return this.latestPoiBlockHash;
+    this.isSyncing = false;
+  }
+
+  private async setLatestSyncedPoi(poiBlock: ProofOfIndex, flush?: boolean): Promise<void> {
+    if (this._latestSyncedPoi !== undefined && this.latestSyncedPoi.id >= poiBlock.id) {
+      throw new Error(
+        `Set latest synced poi out of order, current height ${this.latestSyncedPoi.id}, new height ${poiBlock.id} `
+      );
+    }
+    this._latestSyncedPoi = poiBlock;
+    this.storeCache.metadata.set('latestSyncedPoiHeight', poiBlock.id);
+    this.eventEmitter.emit(PoiEvent.LatestSyncedPoi, {
+      height: poiBlock.id,
+      timestamp: Date.now(),
+    });
+    if (flush) {
+      await this.storeCache.flushCache(true);
+    }
   }
 
-  setLatestPoiBlockHash(hash: Uint8Array): void {
-    this.latestPoiBlockHash = hash;
+  private async createGenesisPoi(genesisPoi: ProofOfIndex): Promise<void> {
+    const poiBlock = PoiBlock.create(
+      genesisPoi.id,
+      genesisPoi.chainBlockHash,
+      genesisPoi.operationHashRoot,
+      this.projectId,
+      GENESIS_PARENT_HASH
+    );
+    this.poiRepo.bulkUpsert([poiBlock]);
+    await this.setLatestSyncedPoi(poiBlock, true);
+    logger.info(`Genesis Poi created at height ${poiBlock.id}!`);
+  }
+
+  private async syncPoiJob(poiBlocks: ProofOfIndex[], logging?: boolean): Promise<void> {
+    const appendedBlocks: ProofOfIndex[] = [];
+    for (let i = 0; i < poiBlocks.length; i++) {
+      const nextBlock = poiBlocks[i];
+      if (this.latestSyncedPoi.id >= nextBlock.id) {
+        throw new Error(
+          `Sync poi block out of order, latest synced poi height ${this.latestSyncedPoi.id}, next poi height ${nextBlock.id}`
+        );
+      }
+      if (this.latestSyncedPoi.id + 1 < nextBlock.id) {
+        // Fill the with default block
+        this.addDefaultPoiBlocks(nextBlock.id - 1, appendedBlocks);
+      }
+      const syncedPoiBlock = PoiBlock.create(
+        nextBlock.id,
+        nextBlock.chainBlockHash,
+        nextBlock.operationHashRoot,
+        this.projectId,
+        this.latestSyncedPoi.hash
+      );
+      appendedBlocks.push(syncedPoiBlock);
+      await this.setLatestSyncedPoi(syncedPoiBlock);
+    }
+    if (appendedBlocks.length) {
+      if (logging) {
+        syncingMsg(appendedBlocks[0].id, appendedBlocks[appendedBlocks.length - 1].id, appendedBlocks.length);
+      }
+      this.poiRepo?.bulkUpsert(appendedBlocks);
+    }
+  }
+
+  private addDefaultPoiBlocks(endHeight: number, appendedBlocks: ProofOfIndex[]): void {
+    for (let i = this.latestSyncedPoi.id + 1; i <= endHeight; i++) {
+      const syncedPoiBlock = PoiBlock.create(i, null, null, this.projectId, this.latestSyncedPoi.hash);
+      appendedBlocks.push(syncedPoiBlock);
+      void this.setLatestSyncedPoi(syncedPoiBlock);
```

Comment:
Why is this void?

Reply:
the `setLatestSyncedPoi ` provide any option to force flush the cache, so we only await if we need to force flush.

When we create default poi blocks with a continuous range height, we can just use void here.

Reply:
Could we just set it at the end of the for loop?

Reply:
it get parentHash from `this.latestSyncedPoi`, so it need to be set at every single block